services/heartbeat/etcd/etcd.go

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package etcd
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/m3db/m3cluster/services/heartbeat"
+	"golang.org/x/net/context"
+)
+
+const heartbeatKeyPrefix = "_hb"
+
+// NewStore creates a heartbeat store based on etcd
+func NewStore(c *clientv3.Client) heartbeat.Store {
+	return &client{l: c.Lease, kv: c.KV}
+}
+
+type client struct {
+	l  clientv3.Lease
+	kv clientv3.KV
+}
+
+func (c *client) Heartbeat(service, id string, ttl time.Duration) error {
+	resp, err := c.l.Grant(context.Background(), int64(ttl/time.Second))
+	if err != nil {
+		return err
+	}
+
+	_, err = c.kv.Put(context.Background(), heartbeatKey(service, id), "", clientv3.WithLease(resp.ID))
```

Comment:
how expensive is the grant + put?

I didn't do that because I want to allow different ttl for each heartbeat and I can't have multiple keys share the same lease, so I didn't think there is too much value keeping track of the lease id and do keepalive on it unless it's really expensive to do the grant+put way.